_posts/blog/2016-09-06-model-view-viewmodel.md

```diff
@@ -0,0 +1,181 @@
+---
+layout: post
+title: "Android Architecture Patterns Part 3:<br/>Model-View-ViewModel"
+description: Our choice for the upday app - the Model-View-ViewModel pattern.
+modified:
+categories: blog
+author: florina_muntenescu
+excerpt: Our choice for the upday app - the Model-View-ViewModel pattern. Find out what it is, how we applied it in Android and why we consider it perfect for us.
+tags: [Android, Architecture, MVVM]
+date: 2016-09-04T15:39:55-04:00
+---
+After four different designs in the first six months of the development of the upday app, we learned one important lesson: we need an architecture pattern that allows fast reaction to design changes! The solution we chose in the end was Model-View-ViewModel. Discover with me what MVVM is; how we are applying it at upday and what makes it so perfect for us.
+
+## The Model-View-ViewModel Pattern
+
+The main players in the MVVM pattern are:
+
+* The *View* - that informs the ViewModel about the user's actions
+* The *ViewModel* - exposes streams of data relevant to the View
+* The *DataModel* - abstracts the data source. The ViewModel works with the DataModel to get and save the data.  
+
+At a first glance, MVVM seems very similar to the Model-View-Presenter pattern, because both of them do a great job in abstracting the view's state and behavior. The Presentation Model abstracts a view independent from a specific user-interface platform whereas the MVVM pattern was created by Microsoft to simplify the **event driven** programming of user interfaces.
+
+If, in the MVP pattern, the Presenter was telling the View directly what to display, in MVVM, **ViewModel exposes streams of events** to which the Views can bind to. Also, the Views notify the ViewModel about different actions. Thus, the MVVM pattern supports two-way data binding between the View and ViewModel and there is a many-to-one relationship between View and ViewModel. View has a reference to ViewModel but **ViewModel has no information about the View**.
```

Comment:
Arguably overuse of commas in the first sentence. 
Perhaps -
> If the MVP Pattern meant that the Presenter was telling the View directly what to display,

Feel free to replace "meant"  with something better.